### ISSUE TITLE
FR: Integrate with animator

### ISSUE THREAD

**Issue by RomkeVdMeulen** [aurelia/dialog#188]
**I'm submitting a feature request**
- **Library Version:**
  1.0.0-beta.1.1.0

**Please tell us about your environment:**
- **Operating System:**
  Linux (Ubuntu 16.04)
- **Node Version:**
  6.2.2
- **NPM Version:**
  3.9.5
- **JSPM OR Webpack AND Version**
  JSPM 0.16.29
- **Browser:**
  Chrome 51
- **Language:**
  TypeScript 1.8.9

**Current behavior:**
If one or more transitions are defined on the dialog container, then the renderer will [wait for the first one to finish](https://github.com/aurelia/dialog/blob/1.0.0-beta.2.0.0/src/dialog-renderer.js#L117) before showing a dialog. Conversely, when the last dialog is closed, the renderer will [wait for the first transition to finish](https://github.com/aurelia/dialog/blob/1.0.0-beta.2.0.0/src/dialog-renderer.js#L154) on the dialog container before removing the dialog overlay and container from the DOM.

**What is the expected behavior?**
Ideally, all animation for the opening and closing of dialogs should be up to the user (with a nice looking default perhaps), and the renderer should await all animations that occur during the opening or closing of dialogs.

In the case of the `CssAnimator` for example, one might expect the renderer to wait for the closing animation before removing the dialog and dialog container from the DOM:

``` html
<template>
  <ai-dialog class="au-animate">
    ...
  </ai-dialog>
</template>
```

``` css
@keyframes fadeOut {
    0% { opacity: 1; }
    100% { opacity: 0; }
}
ai-dialog.au-animate.leave-active {
  animation: fadeOut 2s ease-in-out;
}
```

**What is the motivation / use case for changing the behavior?**
Awaiting animations after manipulating nodes is already [standard behavior in templating](https://github.com/aurelia/templating/blob/1.0.0-rc.1.0.1/src/view-slot.js#L299).

**Comment by EisenbergEffect** [comment]
Absolutely. @jedd-ahyoung Responsibility for animation in the dialog was only a temporary hack that should have been removed some time ago. We need to switch the dialog to be based on the same animation abstraction as the rest of Aurelia. Should only be a matter of using the ViewSlot to add and remove the dialog views. The Animator should take over from there. When you call `add` or `remove` you would just need to check to see if there's a promise returned from those methods. If there is, that indicates an animation was found and is playing. You would need to wait on the promise.

**Comment by RomkeVdMeulen** [comment]
In the mean time I've made a partial workaround in the form of a decorator for Dialog ViewModels using the fact that the renderer does await any promise returned by the `deactive` hook. My implementation relies on [aurelia-property-injection](https://github.com/devaur/aurelia-property-injection) so it needs some rewriting to work with normal constructor DI.

```ts
import {CssAnimator} from "aurelia-animator-css";

export function animated(target?: Function): any {
	let animate = function(dialogConstructor: Function) {
		let proto = dialogConstructor.prototype;
		if (!proto.constructor.injectProperties) {
			proto.constructor.injectProperties = Object.create(null);
		}
		proto.constructor.injectProperties['animator'] = CssAnimator;
		proto.constructor.injectProperties['element'] = Element;

		let attached: Function = proto.hasOwnProperty("attached")
			? proto.attached
			: () => {};
		let attachedWrapper = function(): Promise<any> {
			let dialogWrapper: HTMLElement = this.element.parentElement;
			if (!dialogWrapper) {
				return attached.apply(this);
			}
			return this.animator.enter(dialogWrapper)
				.then(() => attached.apply(this));
		}
		proto.attached = attachedWrapper;

		let deactivate: Function = proto.hasOwnProperty("deactivate")
			? proto.deactivate
			: () => {};
		let deactivateWrapper = function(): Promise<any> {
			let result = deactivate.apply(this);
			let dialogWrapper: HTMLElement = this.element.parentElement;
			if (!dialogWrapper) {
				return result;
			}
			return Promise.resolve(result)
				.then(() => this.animator.leave(dialogWrapper));
		}
		proto.deactivate = deactivateWrapper;
	};

	return target ? animate(target) : animate;
}
```

**Comment by EisenbergEffect** [comment]
@RomkeVdMeulen Would would you be interested in submitting a PR to fix up the real implementation? It shouldn't be too hard since the ViewSlot mechanism already can handle animation on add/remove.

**Comment by RomkeVdMeulen** [comment]
I'll take a look when I have a moment. Perhaps wednesday.

**Comment by simonfox** [comment]
I've started to have a bit of a look at this.

Should be relatively simple to change `hideDialog` to use the `ViewSlot` mechanism.

There is a bit more of a refactoring required for `showDialog` and the `DialogService`. In https://github.com/aurelia/dialog/blob/master/src/dialog-service.js#L55-L73 attaching of the dialog view to the renderers `ViewSlot` occurs via `CompositionEngine.compose`. However the slot hasn't actually been attached the DOM at that point...that happens after the compose promise is resolved in `DialogRenderer.showDialog`. I think the compose call should become the responsibility of the renderer, and should happen after the slot has actually been inserted into the DOM.

Does that sound correct @EisenbergEffect ? I can put together a PR, but a little guidance to ensure I'm on the right track would be awesome.

**Comment by PWKad** [comment]
@simonfox FWIW any PR as a starting point is easier to comment on.  If you can submit a PR here, anything at all really even pseudo-code if you can't get the implementation perfect with questions asking what needs to be done to make X happen we can probably provide feedback.

**Comment by simonfox** [comment]
Cool thanks @PWKad I'm flat out this week but will get something started next week.

**Comment by mttmccb** [comment]
I'm having a look at this, not sure how far I'll get, trying to unpick it, I think refactoring first might be a better option. I would be nice to have a bit more control of the dialogs.

**Comment by StrahilKazlachev** [comment]
So I have 2 crude PoC working. It will take me some more time to get the internal refactoring done so I can present it.
In the mean time some feedback will be appreciated:
1. What should be animated?
    - only the dialog - the provided view.
    - also the elements that the `Renderer` creates e.g, the overlay.
2. If we also want to animate the elements that the `Renderer` creates how do we mark them?
    - they are out of reach for the one defining the dialog view so `au-animate` can't *just* be added.
    - the `Renderer` is pluggable and that means the elements it creates can and probably will differ - that's part of the idea for pluggable `Renderer`.

**Comment by StrahilKazlachev** [comment]
PR open for discussion #326
